### PR TITLE
post: ng v19

### PR DESCRIPTION
**Pull request by cexbrayat**

### CODE REVIEW

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 1]
```diff
@@ -0,0 +1,948 @@
...
+Finally, you can run the `inputs` one,
+which is the most complex and may require manual intervention.
+
+```
+ng generate @angular/core:signals --migrations=outputs --defaults
+ng generate @angular/core:signals --migrations=queries --defaults
+ng generate @angular/core:signals --migrations=inputs --defaults
+```
+
+For each one, you'll get statistics on how many fields have been migrated and how many have been skipped.
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:445]: fields -> properties

Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:445]: That has already been set before. I'd drop the sentence.

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 2]
```diff
@@ -0,0 +1,948 @@
...
+You can then lint, build, and test your application to see if everything is still working as expected.
+You can then re-run the migration with `--insert-todos` to see the reasons why some fields have been skipped.
+Then you can re-run the migration with `--best-effort-mode` to try to migrate them anyway.
+
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:449]: Same for these three above sentences.

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 3]
```diff
@@ -0,0 +1,948 @@
...
+The `APP_INITIALIZER` token is now deprecated in favor of `provideAppInitializer`.
+This new function is a more elegant way to provide an initializer to your application.
+
+Before v19, you would provide an initializer like this:
+
+```ts
+{
+ provide: APP_INITIALIZER,
+ useFactory: (service: InitService) => service.init(),
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:460]: To be fair, inject could already be used, no? So I would rewrite it with inject.

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 4]
```diff
@@ -0,0 +1,948 @@
...
+
+As usual, an automatic migration will take care of this for you when running `ng update` (you may have to refactor a bit if you want to have a nice function with `inject` as I used in the example above).
+
+## Templates
+
+The `@let` syntax, introduced in [Angular v18.1](/2024/07/10/what-is-new-angular-18.1/), is now stable.
+
+A new feature has been added to the templating syntax,
+with the support of the `typeof` operator,
+so you can now write something like `@if (typeof user === 'object') {` in your templates.
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:482]: I would rephrase this sentence as

Expressions in the template can now use the `typeof`operator, so you can write things like

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 5]
```diff
@@ -0,0 +1,948 @@
...
+Then in a child component, you can get the data via DI and the token `ROUTER_OUTLET_DATA`:
+
+```ts
+readonly userModel = inject<Signal<UserModel>>(ROUTER_OUTLET_DATA);
+```
+
+Note that, for the first time I believe, the value you get via DI is not a static value, but a signal!
+That means that every time `userModel` changes in the parent component,
+the signal in the child component will be updated as well.
+Its type is `Signal<unknown>`, so you'll have to cast it to the correct type.
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:518]: I would remove that sentence, because the line of code above makes it clearer (and doesn't really use a cast).

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 6]
```diff
@@ -0,0 +1,948 @@
...
+{
+  "applicationMaxAge": "1d6h" // 1 day and 6 hours
+}
+```
+
+Another new feature is the ability to define a `refreshAhead`
+delay for a specific data group. 
+The duration specifies the time ahead of the expiration of a cached resource
+when the Angular service worker should proactively attempt
+to refresh the resources from the network.
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:544]: I would rephrase the above sentence as 

When the time before the expiration of a cached resource is less than this `refreshAhead` delay, Angular refreshes the resource.

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 7]
```diff
@@ -0,0 +1,948 @@
...
+{
+  "dataGroups": [
+ {
+      "name": "api-users",
+      "urls": ["/api/users/**"],
+      "cacheConfig": {
+        "maxAge": "1d",
+        "timeout": "10s",
+        "refreshAhead": "10m"
+ }
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:558]: Wrong indentation

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 8]
```diff
@@ -0,0 +1,948 @@
...
+ }
+ ]
+}
+```
+
+## SSR
+
+There are a ton of changes in the Server-Side Rendering (SSR) part of Angular,
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:566]: There are a ton of -> There are tons of

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 9]
```diff
@@ -0,0 +1,948 @@
...
+
+### Event Replay is stable
+
+The event replay feature, introduced in [Angular v18](/2024/05/22/what-is-new-angular-18.0/), is now stable.
+The CLI will now generate the necessary `withEventReplay()` call for you
+when you create a new application with SSR.
+
+### Application stability
+
+When working with SSR, you need to know when the application is stable,
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:577]: you need -> Angular needs

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 10]
```diff
@@ -0,0 +1,948 @@
...
+meaning that all the asynchronous operations have been completed,
+in order to render the application to a string and send it to the client.
+Zone.js usually does this but in a zoneless application,
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:580]: usually allows knowing this

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 11]
```diff
@@ -0,0 +1,948 @@
...
+
+```ts
+const userData = await inject(PendingTasks).run(() => fetch('/api/users'));
+//☝️ Angular will wait for the promise to resolve
+this.users.set(usersData);
+```
+
+A new (experimental) RxJS operator called `pendingUntilEvent`
+has also been added to the framework (in the `@angular/core/rxjs-interop` package):
+it allows to mark an observable as important for the application stability until a first event is emitted:
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:603]: it allows marking

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 12]
```diff
@@ -0,0 +1,948 @@
...
+  // Home component
+ { path: '', component: HomeComponent },
+  // About component
+ { path: 'about', component: AboutComponent },
+  // User component with a parameter
+ { path: 'users/:id', component: UserComponent }
+];
+```
+
+In v18, `ng build` generated a `browser` folder with `index.html` and `about/index.html` files (with `"prerender": true`).
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:709]: I'd drop the "(with `"prerender": true`)". We're talking about prerendering, so that's obvious to me.

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 13]
```diff
@@ -0,0 +1,948 @@
...
+
+The server routes configuration doesn't need to define all the routes as you can see,
+with a possible "catch-all" route '**' to define a default behavior.
+If a route doesn't exist though, you'll get an error at build time:
+
+```ts
+✘ [ERROR] The 'unknown' server route does not match any routes defined in the Angular routing configuration (typically provided as a part of the 'provideRouter' call). Please make sure that the mentioned server route is present in the Angular routing configuration.
+```
+
+You can also define a `getPrerenderParams` function, to prerender routes with parameters.
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:761]: The second solution is to define...

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 14]
```diff
@@ -0,0 +1,948 @@
...
+ path: '**',
+ renderMode: RenderMode.Client,
+ }
+];
+```
+
+This file is then loaded in the `app.config.server.ts` file,
+using `provideServerRoutesConfig(serverRoutes)`.
+
+The server routes configuration doesn't need to define all the routes as you can see,
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:753]: I'd drop this, and the following example about unknown routes. It goes into too much details.

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 15]
```diff
@@ -0,0 +1,948 @@
...
+If a route doesn't exist though, you'll get an error at build time:
+
+```ts
+✘ [ERROR] The 'unknown' server route does not match any routes defined in the Angular routing configuration (typically provided as a part of the 'provideRouter' call). Please make sure that the mentioned server route is present in the Angular routing configuration.
+```
+
+You can also define a `getPrerenderParams` function, to prerender routes with parameters.
+
+```ts
+export const serverRoutes: Array<ServerRoute> = [
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:764]: I would either leave the other two routes there (which, according to the above, are necessary), or just display the new route, without the `export const ...`

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 16]
```diff
@@ -0,0 +1,948 @@
...
+ {
+ path: 'users/**',
+ renderMode: RenderMode.Prerender,
+    async getPrerenderParams(): Promise<Array<Record<string, string>>> {
+      // API call to get the user IDs
+      const userIds = await inject(UserService).getUserIds(); 
+      // build an array like [{ id: '1' }, { id: '2' }, { id: '3' }]
+      return userIds.map(id => ({ id }));
+ }
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:773]: wrong indentation

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 17]
```diff
@@ -0,0 +1,948 @@
...
+```ts
+✘ [ERROR] The 'unknown' server route does not match any routes defined in the Angular routing configuration (typically provided as a part of the 'provideRouter' call). Please make sure that the mentioned server route is present in the Angular routing configuration.
+```
+
+You can also define a `getPrerenderParams` function, to prerender routes with parameters.
+
+```ts
+export const serverRoutes: Array<ServerRoute> = [
+ {
+ path: 'users/**',
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:766]: Can't we use `users/:id`?

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 18]
```diff
@@ -0,0 +1,948 @@
...
+      const userIds = await inject(UserService).getUserIds(); 
+      // build an array like [{ id: '1' }, { id: '2' }, { id: '3' }]
+      return userIds.map(id => ({ id }));
+ }
+ },
+```
+
+This will prerender the `users/:id` route for each user ID found by the `UserService`,
+generating `users/1/index.html`, `users/2/index.html`, etc.
+Note that `server.ts` is now used during prerendering,
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:779]: This deserves more than a note. It was a pain in the ass not to be able to use it during development, and a big source of confusion.

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 19]
```diff
@@ -0,0 +1,948 @@
...
+"stylePreprocessorOptions": {
+  "sass": {
+    "silenceDeprecations": ["import"]
+ }
+},
+```
+
+It's also possible to throw an error if a deprecation warning is emitted with `fatalDeprecations`
+and to prepare for future deprecations with `futureDeprecations`.
+This feature is going to be really useful to all Sass users,
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:926]: I would drop this sentence

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 20]
```diff
@@ -0,0 +1,948 @@
...
+as some APIs are getting deprecated and will be removed in Sass v3,
+so you may see a bunch of deprecation warnings appear.
+
+### Strict CSP
+
+A new option has been added to the `ng build` command to enable a
+strict Content Security Policy (CSP) in the generated `index.html` file.
+This option applies the recommendations from the [Web.dev article](https://web.dev/articles/strict-csp#choose-hash)
```
Comment (jnizet) [inline _posts/2024-11-19-what-is-new-angular-19.0.md:934]: from the -> from this ?

**Review by jnizet — CHANGES_REQUESTED** [review]

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 21]
```diff
@@ -0,0 +1,948 @@
...
+
+src/app/user/users.component.ts:9:27:
+  9 │   imports: [UserComponent, UnusedComponent],
+ ╵                            ~~~~~~~~~~~~~
+```
+
+A code action is provided to remove the unused import for you via the language service
+(but no automatic migration).
+
+You can disable the diagnostic if needed with:
```
[the remark this answers is not in the document]
Comment (cexbrayat) [reply _posts/2024-11-19-what-is-new-angular-19.0.md:73]: Yes, exactly. On OiA for example, we had a few hundreds of imports to fix.

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 22]
```diff
@@ -0,0 +1,948 @@
...
+We talk about them in the next section.
+
+`effect` went through some changes and is still in developer preview.
+
+Effects are in fact now of two types:
+effects created in components or directives
+and effects created in root services
+(or in components with the `forceRoot` option), called root effects.
+
+Component effects now run _during_ change detection and not _after_ it as it was the case before.
```
[the remark this answers is not in the document]
Comment (cexbrayat) [reply _posts/2024-11-19-what-is-new-angular-19.0.md:103]: I wanted to point out that they run as part of the change detection, at the beginning of it (hence running `detectChanges` in tests triggers them)

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 23]
```diff
@@ -0,0 +1,948 @@
...
+
+Note that you can use the previous value of the source signal in the computation function if you need to.
+For example, if you want to access the previous items value to compare it with the new one,
+you can declare the `linkedSignal` with the source and computation options.
+In that case, the computation function receives the current and previous values of the source as parameters.
+
+```ts
+export class ItemListComponent {
+ items = input.required<Array<ItemModel>>();
+ selectedItem = linkedSignal</* source */ Array<ItemModel>, /* value */ ItemModel>({
```
[the remark this answers is not in the document]
Comment (cexbrayat) [reply _posts/2024-11-19-what-is-new-angular-19.0.md:235]: Otherwise, previousChoice is not properly inferred. So we either need to cast `previous.value` to `itemModel` or add the generic parameters to the linked signal.

File: _posts/2024-11-19-what-is-new-angular-19.0.md [hunk 24]
```diff
@@ -0,0 +1,948 @@
...
+    return (await response.json()) as Array<UserModel>;
+ }
+});
+```
+
+If the `sortOrder` signal changes,
+the resource will automatically reload!
+You can also cancel the previous request if needed when the resource is reloaded
+using the [`abortSignal`](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) parameter of the loader
+(for example to implement a debounce).
```
[the remark this answers is not in the document]
Comment (cexbrayat) [reply _posts/2024-11-19-what-is-new-angular-19.0.md:342]: FYI a debounce would look like this:
```ts
loader: async params => {
  const abortSignal = params.abortSignal;
  await timeout(400);
  if (abortSignal.aborted) {
    return;
  }
  const response = await fetch(`/users?sort=${request.sort}`);
  // ...
```
But I agree that the abortSignal code should probably always be there.

File: _posts/2024-11-19-what-is-new-angular-19.0.md (hunk 3 above)
Comment (cexbrayat) [reply _posts/2024-11-19-what-is-new-angular-19.0.md:460]: Right. FYI the init function now runs in the injection context, so it can use `inject` directly.

File: _posts/2024-11-19-what-is-new-angular-19.0.md (hunk 19 above)
Comment (cexbrayat) [reply _posts/2024-11-19-what-is-new-angular-19.0.md:926]: wait until you see the pile of warnings we get on our projects 😅
